## OR 506: HomeWork-2 
**Created in iPython Notebook by Guduguntla Vamshi <gudugu@ncsu.edu>**

**1.** Approximate the minimum value of the function : 
\begin{equation}
  f(x) = exp(-x) + x^2
\end{equation}


**i.Golden Section Method**
Results:

In [8]:
Golden_section(-100,100,.001)

Starting with Upper and Lower Bound  100 -100
100.0 -23.6068
52.7864 -23.6068
23.6068 -23.6068
23.6068 -5.57281
12.46118 -5.57281
5.57281 -5.57281
5.57281 -1.31556
2.94169 -1.31556
1.31556 -1.31556
1.31556 -0.31056
0.69444 -0.31056
0.69444 0.07331
0.45719 0.07331
0.45719 0.21994
0.45719 0.31056
0.40118 0.31056
0.36657 0.31056
0.36657 0.33195
0.36657 0.34518
0.3584 0.34518
0.35335 0.34518
0.35335 0.3483
0.35335 0.35023
0.35216 0.35023
0.35216 0.35096
0.35216 0.35142

Converged after 26 iterations 
Point is :  0.351787


Comments on Result:
- The tolerance for the convergence is set to be **0.001**
- The upper bound is **100** and the lower bound is **-100**
- Using the above values, the golden section implementation converged after **26** iterations.
- Therefore, the point of minima is **0.351787**

### Algorithm Implementation 

In [1]:
# Import Libraries and setting the Golden ratio value
import math
R  = ((5 ** 0.5) - 1) / 2
import numpy as np

In [2]:
# function definition
def fun(x): return math.exp(-x) + x**2

#method to compute the minima by golden section 
def Golden_section(lb,ub,tol):
    iteration = 0          # counter to count the number of iterations 
    x1 = ub - ((ub-lb)*R)  # setting the x1 point by  golden ratio value property
    x2 = lb + ((ub-lb)*R)  # setting the x2 point by  golden ratio value property
    f1 = fun(x1)           
    f2 = fun(x2)
    print "Starting with Upper and Lower Bound ",ub,lb
    while (abs(ub-lb) > tol): #loop to check for the tolerance
        iteration += 1       
        if f1 < f2:           #condition to check for the location of minima, to the left
            ub = x2
            print round(ub,5),round(lb,5)
            x2 = x1           # set the values accordingly
            f2 = f1
            x1 = ub - ((ub-lb)*R)
            f1 = fun(x1)
        else:
            lb = x1           #condition otherwise
            print round(ub,5),round(lb,5)
            x1 = x2
            f1 = f2
            x2 = lb + ((ub-lb)*R)
            f2 = fun(x2)

    print "\nConverged after %g iterations \nPoint is : "%iteration,
                                                    round(.5*(ub+lb),6)

**ii.BiSection Method**
Results:

In [9]:
Bi_section(-100,100,.001)

Starting with Upper and Lower Bound  100 -100
100.0 0.0
50.0 0.0
25.0 0.0
12.5 0.0
6.25 0.0
3.125 0.0
1.5625 0.0
0.78125 0.0
0.39063 0.0
0.39063 0.19531
0.39063 0.29297
0.39063 0.3418
0.36621 0.3418
0.354 0.3418
0.354 0.3479
0.354 0.35095
0.35248 0.35095
0.35248 0.35172

Converged after 18 iterations 
Point is :  0.352097


Comments on Result:
- The tolerance for the convergence is set to be **0.001**
- The upper bound is **100** and the lower bound is **-100**
- Using the above values, the golden section implementation converged after **18** iterations.
- Therefore, the point of minima is **0.352097**

### Algorithm Implementation 

In [4]:
def fun(x): return math.exp(-x) + x**2 
def fun_derv(x): return -math.exp(-x) + 2*x  #function to return the derivative
def Bi_section(lb,ub,tol): #method to compute the minima by bisection 
    iteration = 0
    c = .5*(lb+ub)
    fc = fun_derv(c)
    print "Starting with Upper and Lower Bound ",ub,lb
    while (abs(ub-lb) > tol):
        iteration += 1 # counter to count the number of iterations 
        if fc < 0: #check for the derivative of the point whether negative
            lb = c     # setting the lb point by golden bisection method
            print round(ub,5),round(lb,5)
            c = .5*(lb+ub) # setting the c point by taking the mean
            fc = fun_derv(c) #check for the derivative
        else:
            ub = c
            print round(ub,5),round(lb,5)
            c = .5*(lb+ub)
            fc = fun_derv(c)

    print "\nConverged after %g iterations \nPoint is : "
                                    %iteration,round(.5*(ub+lb),6)

**2.i** Newtons Method : 
\begin{equation}
  f(x) =7x^4 + 3x^3 + 2x^2 + 9x + 4 = 0
\end{equation}

**i.Newtons Method**
Results:

Comments on Result:
- The tolerance for the convergence is set to be **0.000001**
- Using the initial guess as **-1** approxiamated value of x is **-0.909306388848**
- Using the initial guess as **1** approxiamated value of x is **-0.511041787678**
- Since it is the order four equation, and with other trials, I found that the method converges to these two values only. Hence, the other two roots are **complex**
- The computed values are cross-validates by checking plugging in these values and and checking for convergence ~0.

In [15]:
x_approx = -1  # rough guess
x = newtons_method(f, x_approx) #compute
print("Solve for x in 7*x**4 + 3*x**3 + 2*x**2 + 9*x + 4 = 0")
print("x = %0.12f" % x)
print("Testing with the above x value ...")
print("%0.12f" % (7*x**4 + 3*x**3 + 2*x**2 + 9*x + 4))  

Solve for x in 7*x**4 + 3*x**3 + 2*x**2 + 9*x + 4 = 0
x = -0.909306388848
Testing with the above x value ...
0.000000000487


In [16]:
x_approx = 1  # rough guess
x = newtons_method(f, x_approx)
print("Solve for x in 7*x**4 + 3*x**3 + 2*x**2 + 9*x + 4 = 0")
print("x = %0.12f" % x)
print("Testing with the above x value ...")
print("%0.12f" % (7*x**4 + 3*x**3 + 2*x**2 + 9*x + 4))  

Solve for x in 7*x**4 + 3*x**3 + 2*x**2 + 9*x + 4 = 0
x = -0.511041787678
Testing with the above x value ...
0.000000002000


### Implementation of Newtons method

In [12]:
import numpy as np
import math

def derivative(f): #method to compute the derivative
    def compute(x, dx):
        return (f(x+dx) - f(x))/dx
    return compute

def newtons_method(f, x, dx=0.000001, tolerance=0.000001): #newton method
    df = derivative(f)
    while True:
        x1 = x - f(x)/df(x, dx) # loop to check
        t = abs(x1 - x) # taking the absolute value for convergence
        if t < tolerance:
            break # breaks out the loops to return the value of x 
        x = x1
    return x

def f(x):
    return 7*x**4 + 3*x**3 + 2*x**2 + 9*x + 4 # the function

**2.ii** Newtons Method for solving simultaneous equations: 
\begin{equation}
  f(x,y) = 3xy + 7x + 2y - 3 \\
  g(x,y) = 5xy - 9x - 4y + 6
\end{equation}

Result and Comments:
- The tolerance vector for the convergence is set to be **0.000001**
- Using the initial guess as **[-2,-2]** approxiamated value of x is **[0.693548,-0.454545]**
- Using the initial guess as **[2,2]** approxiamated value of x is **[0.0,1.5]**
- The algorithm converged in **7** iterations using the above guess vector
- The computed values are cross-validates by checking plugging in these values and and checking for convergence ~0.

### Implementation

In [18]:
import numpy as np
import math
from sympy import symbols, diff
def f1(x,y): #function 1
    return 3*x*y + 7*x + 2*y - 3
def f2(x,y): #function 2
    return 5*x*y - 9*x - 4*y + 6
def f1_x(x,y): #derivatives wrt x
    return 3*y + 7
def f1_y(x,y): #derivatives wrt y
    return 3*x + 2
def f2_x(x,y): #derivatives wrt x
    return 5*y - 9
def f2_y(x,y): #derivatives wrt x
    return 5*x - 4
def newton_equation(f1,f2,guess):
    iteration = 0
    while True:
        iteration += 1 #compute the matrix of derivatives in Dx
        Dx = np.matrix([[f1_x(guess.item(0),guess.item(1)),f1_y(guess.item(0),
                                                                guess.item(1))],
                        [f2_x(guess.item(0),guess.item(1)),f2_y(guess.item(0),
                                                                guess.item(1))]])
        Dx_inv = np.linalg.inv(Dx)    # inverse of the derivative matrix Dx               
        function_vector = np.matrix([[f1(guess.item(0),guess.item(1))],
                                     [f2(guess.item(0),guess.item(1))]])
        #function vector
        t = guess - np.dot(Dx_inv,function_vector) # new vector 
        function_vector1 = np.matrix([[f1(t.item(0),t.item(1))],
                                      [f2(t.item(0),t.item(1))]])
        #function value at new vector
        print "Iteration :%g the guess vector is:"%iteration,
                                    round(t.item(0),6),round(t.item(1),6)
        if (abs(function_vector1.item(0)) < 0.00000000001 and 
                                abs(function_vector1.item(1)) < 0.00000000001):
            break
        guess = t #re-initialzing the guess vector to the computed value of current vector
    print "\nThe function value has converged!",round(function_vector1.item(0)),
                                                            round(function_vector1.item(1))
guess = np.matrix([[-2],[-2]])
newton_equation(f1,f2,guess)             

Iteration :1 the guess vector is: 1.711111 -3.322222
Iteration :2 the guess vector is: 1.073013 -1.523944
Iteration :3 the guess vector is: 0.792685 -0.733929
Iteration :4 the guess vector is: 0.704569 -0.485602
Iteration :5 the guess vector is: 0.693718 -0.455024
Iteration :6 the guess vector is: 0.693548 -0.454546
Iteration :7 the guess vector is: 0.693548 -0.454545

The function value has converged! -0.0 -0.0


**using the guess vector as [2,2]**

The other root is **[0,1.5]**